### PR TITLE
Adding postgres config store in conformance tests

### PR DESCRIPTION
**Pull request by shivamkm07**
# Description

This PR adds postgres config store in conformance tests. It also fixes some bugs required in the store implementation and includes the updater implementation for postgres. 

## Issue reference

We strive to have all PR being opened based on an issue, where the problem or feature have been discussed prior to implementation.

Please reference the issue this PR will close: #_[issue number]_

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

* [ ] Code compiles correctly
* [ ] Created/updated tests
* [ ] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_

### CODE REVIEW

File: configuration/postgres/postgres.go
Comment (ItalyPaleAle) [inline configuration/postgres/postgres.go:72]: Why are you lowercasing the table?

File: configuration/postgres/postgres.go
Comment (ItalyPaleAle) [inline configuration/postgres/postgres.go:167]:
```suggestion
		if strings.EqualFold(pgNotifyChannelKey, k) {
```

File: tests/conformance/configuration/configuration.go
Comment (ItalyPaleAle) [inline tests/conformance/configuration/configuration.go:157]:
```suggestion
			assert.NoError(t, err)
```

**Review by ItalyPaleAle — CHANGES_REQUESTED** [review]

File: configuration/postgres/postgres.go
Comment (shivamkm07) [reply configuration/postgres/postgres.go:72]: This is because table names in pg_tables are stored in lowercase only, and SELECT clause does the case sensitive comparison, so otherwise the query output is always false:

image

File: tests/conformance/configuration/configuration.go
Comment (shivamkm07) [reply tests/conformance/configuration/configuration.go]: Done

File: tests/conformance/configuration/configuration.go
Comment (shivamkm07) [reply tests/conformance/configuration/configuration.go]: Done

File: configuration/postgres/postgres.go
Comment (shivamkm07) [reply configuration/postgres/postgres.go]: @ItalyPaleAle Please note that this would be a breaking change in postgres metadata. cc @mukundansundar

File: configuration/postgres/postgres.go
Comment (shivamkm07) [inline configuration/postgres/postgres.go:165]: Removing the list as there could be only one channel passed in metadata in one subscribe call.

File: configuration/postgres/postgres.go
Comment (shivamkm07) [inline configuration/postgres/postgres.go:391]: Making `ActiveSubscriptions` with subscribeIDs as key instead of channel, because there could be multiple subscription IDs for a common channel (multiple subscribe calls with same channel), but there would be a unique subscriptionID for a given subscribe call with a specific channel

File: tests/conformance/configuration/configuration.go
Comment (shivamkm07) [inline tests/conformance/configuration/configuration.go:220]: Ideally Conformance tests should not have component specific code as they test the interface, but subscribe call for postgres required channel in the metadata and so it's added here

File: configuration/postgres/postgres.go
Comment (ItalyPaleAle) [reply configuration/postgres/postgres.go]: It's an alpha component so ok to add breaking changes.

File: .github/scripts/test-info.mjs
Comment (ItalyPaleAle) [inline .github/scripts/test-info.mjs:188]: Please run `make prettier-format`. A bit of a nit but this should go before `configuration.redis.v6` to keep alphabetic order :)

File: configuration/postgres/postgres.go
Comment (ItalyPaleAle) [inline configuration/postgres/postgres.go:169]: Add a `break` after the match to short-circuit

File: configuration/postgres/postgres.go
Comment (ItalyPaleAle) [inline configuration/postgres/postgres.go:168]: Since these keys are all ASCII-only (or anyways we don't care if they're not ASCII), please do:

```suggestion
		if strings.ToLower(k) == pgNotifyChannelKey {
```

The linter will complain so you may need to add `//nolint:gocritic`

> strings.EqualFold is useful when the strings contain non-ASCII chars, but in this case is just slower.

File: configuration/postgres/postgres.go
Comment (ItalyPaleAle) [inline configuration/postgres/postgres.go:186]: Since the constants `unlistenTemplate` and `listenTemplate` are only used once, please inline them. It makes the code a lot more readable...

(Also consider using simple string concatenation: `pgChannel := "UNLISTEN " + sub.channel` )

File: configuration/postgres/postgres.go
Comment (ItalyPaleAle) [inline configuration/postgres/postgres.go:181]: Please short-circuit this instead:

```go
if p.ActiveSubscriptions[req.ID] == "" {
    return fmt.Errorf("unable to find subscription with ID : %v", req.ID)
}

// Rest of the code
```

This too makes the code more readable by reducing the number of indented blocks and by making the exit condition more immediate.

File: tests/conformance/configuration/configuration.go
Comment (ItalyPaleAle) [inline tests/conformance/configuration/configuration.go:152]:
```suggestion
		require.NoError(t, err)
```

File: tests/conformance/configuration/configuration.go
Comment (ItalyPaleAle) [inline tests/conformance/configuration/configuration.go:155]: Please inline the constants here

File: tests/conformance/configuration/configuration.go
Comment (ItalyPaleAle) [inline tests/conformance/configuration/configuration.go:157]:
```suggestion
			require.NoError(t, err)
```

File: tests/conformance/configuration/configuration.go
Comment (ItalyPaleAle) [inline tests/conformance/configuration/configuration.go:164]:
```suggestion
		require.NoError(t, err)
```

File: tests/conformance/configuration/configuration.go
Comment (ItalyPaleAle) [inline tests/conformance/configuration/configuration.go:166]: Here you may also want to add a line that makes the conf test exit if the init failed. If the component couldn't be initialized, there's no point in trying to run the tests.

```go
if t.Failed() {
    t.Fatal("initialization failed")
}
```

File: tests/utils/configupdater/postgres/postgres.go
Comment (ItalyPaleAle) [inline tests/utils/configupdater/postgres/postgres.go:19]: Please inline these constants since they're strings and used just once.

There's one scenario where it makes sense to define errors as global variables, and that's when:

1. They are defined as errors and not strings, so `var ErrMissingConnectionString = errors.New("missing postgreSQL connection string")`
2. This makes it possible for callers that receive an error to then check if the error matches (even if wrapped). So they can do `errors.Is(err, postgres.ErrorMissingConnectionString)`

I don't think this is necessary here so these can just be inlined as strings.

File: configuration/postgres/postgres.go
Comment (ItalyPaleAle) [inline configuration/postgres/postgres.go:367]:
```suggestion
    val := p.ActiveSubscriptions[subscriptionID]
	if val.channel == channel && (len(val.keys) == 0 || slices.Contains(val.keys, key)) {
```

**Review by ItalyPaleAle — CHANGES_REQUESTED** [review]

File: configuration/postgres/postgres.go
Comment (shivamkm07) [inline configuration/postgres/postgres.go:182]: This is to return the `latest` version of the key if there are multiple versions present in the config store as per discussed in https://github.com/dapr/components-contrib/issues/2187
cc @ItalyPaleAle

File: configuration/postgres/postgres.go
Comment (ItalyPaleAle) [inline configuration/postgres/postgres.go:327]: (Not going to block the PR on this) FYI you can use a single regex to validate the charset and length: https://regex101.com/r/hNxNfd/1

**Comment by ItalyPaleAle** [comment]
Hi @shivamkm07 thanks for this PR. I will merge it for now to unblock you, but I'd like to ask that you please make a follow-up PR to address some perf issues.

**First**, in `getUniqueItemsPerKey` you are already iterating through the `res` array once. 

https://github.com/shivamkm07/components-contrib/blob/db91a2b742fae876a871ccebdfb9747927e21074/configuration/postgres/postgres.go

What you want to do is:

1. Initialize `items := make(map[string]*configurationItem)` before the first loop
1. While you iterate through `res` the first time (in `for _, r := range res`):
   - If `items[r.key] == nil` is true, then pick this item as the one for `r.key`
   - If there's already an item,  in `items[r.key]`, right there you check if `items[r.key].Version < r.Version`. If yes, then you replace `items[r.key]` with this item

This will allow both having a single iteration over the loop, and saves allocating a lot of temporary memory.

**Second** when you compare versions in `compareVersion`, every time you're converting from string to int:

https://github.com/shivamkm07/components-contrib/blob/db91a2b742fae876a871ccebdfb9747927e21074/configuration/postgres/postgres.go

`strconv.Atoi` is fast, but not free.

So when you iterate through the `res` array, you should also have a `map[string]int` where you store the integer value of the the latest version of the item. So then you don't have to re-convert the same version from string to int on each iteration.

**Review by berndverst — DISMISSED** [review]
@ItalyPaleAle please do not merge until perf improvements have been made in *this* PR. Follow up PRs never get done in a timely manner and I don't see a reason for doing so.

**Comment by shivamkm07** [comment]
> Hi @shivamkm07 thanks for this PR. I will merge it for now to unblock you, but I'd like to ask that you please make a follow-up PR to address some perf issues.
> 
> **First**, in `getUniqueItemsPerKey` you are already iterating through the `res` array once.
> 
> https://github.com/shivamkm07/components-contrib/blob/db91a2b742fae876a871ccebdfb9747927e21074/configuration/postgres/postgres.go
> 
> What you want to do is:
> 
> 1. Initialize `items := make(map[string]*configurationItem)` before the first loop
> 2. While you iterate through `res` the first time (in `for _, r := range res`):
>    
>    * If `items[r.key] == nil` is true, then pick this item as the one for `r.key`
>    * If there's already an item,  in `items[r.key]`, right there you check if `items[r.key].Version < r.Version`. If yes, then you replace `items[r.key]` with this item
> 
> This will allow both having a single iteration over the loop, and saves allocating a lot of temporary memory.
> 
> **Second** when you compare versions in `compareVersion`, every time you're converting from string to int:
> 
> https://github.com/shivamkm07/components-contrib/blob/db91a2b742fae876a871ccebdfb9747927e21074/configuration/postgres/postgres.go
> 
> `strconv.Atoi` is fast, but not free.
> 
> So when you iterate through the `res` array, you should also have a `map[string]int` where you store the integer value of the the latest version of the item. So then you don't have to re-convert the same version from string to int on each iteration.

Done

**Review by ItalyPaleAle — APPROVED** [review]

**Pull request merged**
